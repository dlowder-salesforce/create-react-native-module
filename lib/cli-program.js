const program = require('commander');
const updateNotifier = require('update-notifier');

const command = require('./cli-command');
const pkg = require('./../package.json');

updateNotifier({ pkg }).notify();

program
  .version(pkg.version)
  .usage(command.usage)
  .description(command.description)
  .action(function runAction () {
    if (!arguments.length) return program.help();
    command.func(arguments, {}, this.opts());
  });

(command.options || [])
  .forEach(opt => program.option(
    opt.command,
    opt.description,
    opt.parse || (value => value),
    opt.default
  ));

program.parse(process.argv);
