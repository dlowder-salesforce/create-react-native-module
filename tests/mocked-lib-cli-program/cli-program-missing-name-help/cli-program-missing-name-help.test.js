// special compact mocks for this test:
const mysnap = [];
const mockpushit = x => mysnap.push(x);
jest.mock('update-notifier', () => ({ pkg }) => {
  // only check a limited number of fields in pkg
  expect(pkg.name).toBeDefined();
  expect(pkg.version).toBeDefined();
  return { notify: () => mockpushit({ notify: {} }) };
});
jest.mock('fs-extra', () => ({
  outputFile: (outputFileName, theContent) => {
    mockpushit({ outputFileName, theContent });
    return Promise.resolve();
  },
  ensureDir: (dir) => {
    mockpushit({ ensureDir: dir });
    return Promise.resolve();
  },
}));
const mockCommanderState = { actionFunction: null };
const mockCommander = {
  args: [],
  version: (version) => {
    // actual value of version not expected to be stable
    expect(version).toBeDefined();
    mockpushit({ version: 'x' });
    return mockCommander;
  },
  usage: (usage) => {
    mockpushit({ usage });
    return mockCommander;
  },
  description: (description) => {
    mockpushit({ description });
    return mockCommander;
  },
  action: (actionFunction) => {
    mockpushit({ action: { actionFunction } });
    mockCommanderState.actionFunction = actionFunction;
    return mockCommander;
  },
  option: (...args) => {
    mockpushit({ option: { args } });
    return mockCommander;
  },
  parse: (argv) => {
    mockpushit({ parse: { argv } });
    mockCommanderState.actionFunction.apply(
      { opts: () => ({ platforms: 'android,ios' }) },
      []);
  },
  help: () => {
    mockpushit({ help: {} });
  },
};
jest.mock('commander', () => mockCommander);

// TBD hackish mock(s):
process.argv = ['node', 'create-cli.js'];

test('mocked cli-program.js runs correctly defaults', () => {
  require('../../../lib/cli-program.js');
  // Using a 1 ms timer to wait for the
  // CLI program func to finish.
  // FUTURE TBD this looks like a bad smell
  // that should be resolved someday.
  return new Promise((resolve) => setTimeout(resolve, 1))
    .then(() => { expect(mysnap).toMatchSnapshot(); });
});
