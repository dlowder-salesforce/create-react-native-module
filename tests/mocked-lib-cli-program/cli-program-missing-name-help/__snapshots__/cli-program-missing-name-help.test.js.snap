// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mocked cli-program.js runs correctly defaults 1`] = `
Array [
  Object {
    "notify": Object {},
  },
  Object {
    "version": "x",
  },
  Object {
    "usage": "[options] <name>",
  },
  Object {
    "description": "creates a React Native library module for one or more platforms",
  },
  Object {
    "action": Object {
      "actionFunction": [Function],
    },
  },
  Object {
    "option": Object {
      "args": Array [
        "--prefix [prefix]",
        "The prefix for the library module",
        [Function],
        "",
      ],
    },
  },
  Object {
    "option": Object {
      "args": Array [
        "--module-name [moduleName]",
        "The module library package name to be used in package.json. Default: react-native-(name in param-case)",
        [Function],
        undefined,
      ],
    },
  },
  Object {
    "option": Object {
      "args": Array [
        "--module-prefix [modulePrefix]",
        "The module prefix for the library module, ignored if --module-name is specified",
        [Function],
        "react-native",
      ],
    },
  },
  Object {
    "option": Object {
      "args": Array [
        "--package-identifier [packageIdentifier]",
        "(Android only!) The package name for the Android module",
        [Function],
        "com.reactlibrary",
      ],
    },
  },
  Object {
    "option": Object {
      "args": Array [
        "--platforms <platforms>",
        "Platforms the library module will be created for - comma separated",
        [Function],
        "ios,android",
      ],
    },
  },
  Object {
    "option": Object {
      "args": Array [
        "--github-account [githubAccount]",
        "The github account where the library module is hosted",
        [Function],
        "github_account",
      ],
    },
  },
  Object {
    "option": Object {
      "args": Array [
        "--author-name [authorName]",
        "The author's name",
        [Function],
        "Your Name",
      ],
    },
  },
  Object {
    "option": Object {
      "args": Array [
        "--author-email [authorEmail]",
        "The author's email",
        [Function],
        "yourname@email.com",
      ],
    },
  },
  Object {
    "option": Object {
      "args": Array [
        "--license [license]",
        "The license type",
        [Function],
        "MIT",
      ],
    },
  },
  Object {
    "option": Object {
      "args": Array [
        "--view",
        "Generate the module as a very simple native view component",
        [Function],
        undefined,
      ],
    },
  },
  Object {
    "option": Object {
      "args": Array [
        "--use-cocoapods",
        "Generate a library with a sample podspec and third party pod usage example",
        [Function],
        undefined,
      ],
    },
  },
  Object {
    "option": Object {
      "args": Array [
        "--generate-example",
        "Generate an example project and links the library module to it, requires both react-native-cli and yarn to be installed globally",
        [Function],
        undefined,
      ],
    },
  },
  Object {
    "option": Object {
      "args": Array [
        "--example-name [exampleName]",
        "Name for the example project",
        [Function],
        "example",
      ],
    },
  },
  Object {
    "option": Object {
      "args": Array [
        "--example-react-native-version [exampleReactNativeVersion]",
        "React Native version for the generated example project",
        [Function],
        "react-native@0.59",
      ],
    },
  },
  Object {
    "parse": Object {
      "argv": Array [
        "node",
        "create-cli.js",
      ],
    },
  },
  Object {
    "help": Object {},
  },
]
`;
